File: record.go
package main

import (
	"bytes"
	"context"
	"encoding/json"
	"errors"
	"io"
	"net/http"
	"os"
	"os/exec"
	"path/filepath"
	"strconv"
	"strings"
	"time"
)

// Walk only the requested JSON path. Decoder offsets refer to the original bytes,
// so indentation, escaped strings and identical values elsewhere cannot shift it.
func jsonOffset(raw []byte, path []string) (int, error) {
	if len(path) == 0 {
		return len(raw) - len(bytes.TrimLeft(raw, " \t\r\n")), nil
	}
	decoder := json.NewDecoder(bytes.NewReader(raw))
	token, err := decoder.Token()
	if err != nil || (token != json.Delim('{') && token != json.Delim('[')) {
		return 0, errors.New("JSON path not found")
	}
	for index := 0; decoder.More(); index++ {
		key := strconv.Itoa(index)
		if token == json.Delim('{') {
			name, err := decoder.Token()
			if err != nil {
				return 0, err
			}
			key = name.(string)
		}
		var value json.RawMessage
		if err := decoder.Decode(&value); err != nil {
			return 0, err
		}
		if key == path[0] {
			offset, err := jsonOffset(value, path[1:])
			return int(decoder.InputOffset()) - len(value) + offset, err
		}
	}
	return 0, errors.New("JSON path not found")
}

type recordLocation struct {
	Path string `json:"path"`
	Line int    `json:"line"`
}

func locateRecord(directory, id, eventID, field string) (recordLocation, error) {
	var location recordLocation
	if !eventIDPattern.MatchString(eventID) || (field != "" && !strings.HasPrefix(field, "/input") && !strings.HasPrefix(field, "/output")) {
		return location, errors.New("invalid event or field")
	}
	var segments []string
	if field != "" {
		for _, segment := range strings.Split(field[1:], "/") {
			segments = append(segments, strings.ReplaceAll(strings.ReplaceAll(segment, "~1", "/"), "~0", "~"))
		}
		if segments[0] != "input" && segments[0] != "output" {
			return location, errors.New("invalid field")
		}
	}
	raw, err := readRunFile(directory, id, "run.json")
	if err == nil {
		run, err := decodeHistory(raw, id)
		if err != nil {
			return location, err
		}
		for index, event := range run.Events {
			if event.ID != eventID {
				continue
			}
			offset, err := jsonOffset(raw, append([]string{"events", strconv.Itoa(index)}, segments...))
			if err != nil {
				return location, err
			}
			return recordLocation{filepath.Join(directory, id, "run.json"), bytes.Count(raw[:offset], []byte("\n")) + 1}, nil
		}
		return location, errors.New("event not found")
	}
	if !os.IsNotExist(err) {
		return location, err
	}
	raw, err = readRunFile(directory, id, "trace.jsonl")
	if err != nil {
		return location, err
	}
	// During a run, prefer the latest complete start/finish record for this event.
	// An unfinished last line is ignored while the recorder appends to the file.
	for index, line := range bytes.Split(raw, []byte("\n")) {
		var record struct {
			ID    string `json:"id"`
			Phase string `json:"phase"`
		}
		if json.Unmarshal(line, &record) != nil || record.ID != eventID || (record.Phase != "start" && record.Phase != "finish") {
			continue
		}
		if _, err := jsonOffset(line, segments); err == nil {
			location = recordLocation{filepath.Join(directory, id, "trace.jsonl"), index + 1}
		}
	}
	if location.Line == 0 {
		return location, errors.New("event not yet saved")
	}
	return location, nil
}

func recordEditorCommand(ctx context.Context, location recordLocation) (*exec.Cmd, string, error) {
	// Prefer the installed app CLI: a same-named terminal agent may shadow it in PATH.
	for _, candidate := range []struct{ path, name string }{
		{"/Applications/Cursor.app/Contents/Resources/app/bin/cursor", "Cursor"},
		{"/Applications/Visual Studio Code.app/Contents/Resources/app/bin/code", "VS Code"},
		{"cursor", "Cursor"}, {"code", "VS Code"},
	} {
		if executable, err := exec.LookPath(candidate.path); err == nil {
			return exec.CommandContext(ctx, executable, "--goto", location.Path+":"+strconv.Itoa(location.Line)), candidate.name, nil
		}
	}
	return nil, "", errors.New("未找到 Cursor 或 VS Code；请安装编辑器及其 cursor / code 命令")
}

func (s *Server) openRecord(w http.ResponseWriter, r *http.Request, id string) {
	var payload struct {
		EventID string `json:"event_id"`
		Field   string `json:"field"`
	}
	decoder := json.NewDecoder(http.MaxBytesReader(w, r.Body, 4096))
	decoder.DisallowUnknownFields()
	var extra any
	if decoder.Decode(&payload) != nil || decoder.Decode(&extra) != io.EOF {
		respond(w, 400, map[string]any{"error": "记录定位参数无效"})
		return
	}
	location, err := locateRecord(s.state, id, payload.EventID, payload.Field)
	if err != nil {
		respond(w, 404, map[string]any{"error": "未找到对应的本地记录或字段；请稍后重试"})
		return
	}
	ctx, cancel := context.WithTimeout(r.Context(), 10*time.Second)
	defer cancel()
	command, editor, err := recordEditorCommand(ctx, location)
	if err == nil {
		err = command.Run()
		if err != nil {
			err = errors.New("编辑器未能打开记录；请检查本机 cursor / code 命令")
		}
	}
	if err != nil {
		respond(w, 500, map[string]any{"error": err.Error()})
		return
	}
	respond(w, 200, struct {
		recordLocation
		Editor string `json:"editor"`
	}{location, editor})
}

File: src/record.ts
import { spawn } from 'node:child_process';
import { accessSync, constants } from 'node:fs';
import { delimiter, join } from 'node:path';
import ts from 'typescript';
import { decodeHistory, eventIDPattern, object, readRunFile } from './storage.ts';
// Use the installed TS parser's JSON AST for positions in original text, not reserialized JSON.
export function jsonOffset(raw: string, path: string[]): number {
    JSON.parse(raw);
    const tree = ts.parseJsonText('record.json', raw);
    const statement = tree.statements[0];
    if (!statement || !ts.isExpressionStatement(statement))
        throw new Error('JSON path not found');
    let node: ts.Node = statement.expression;
    for (const segment of path) {
        if (ts.isObjectLiteralExpression(node)) {
            const property = node.properties.find(p => ts.isPropertyAssignment(p) && ts.isStringLiteral(p.name) && p.name.text === segment);
            if (!property || !ts.isPropertyAssignment(property))
                throw new Error('JSON path not found');
            node = property.initializer;
        }
        else if (ts.isArrayLiteralExpression(node) && /^(0|[1-9][0-9]*)$/.test(segment) && Number(segment) < node.elements.length)
            node = node.elements[Number(segment)];
        else
            throw new Error('JSON path not found');
    }
    return node.getStart(tree);
}
export function locateRecord(directory: string, id: string, eventID: string, field: string): {
    path: string;
    line: number;
} {
    if (!eventIDPattern.test(eventID) || field && !/^\/(input|output)(\/|$)/.test(field))
        throw new Error('invalid event or field');
    const segments = field ? field.slice(1).split('/').map(s => s.replace(/~1/g, '/').replace(/~0/g, '~')) : [];
    let raw: Buffer;
    try {
        raw = readRunFile(directory, id, 'run.json');
    }
    catch (error) {
        if ((error as NodeJS.ErrnoException).code !== 'ENOENT')
            throw error;
        const trace = readRunFile(directory, id, 'trace.jsonl').toString();
        let line = 0;
        trace.split('\n').forEach((raw, index) => {
            try {
                const record = object(JSON.parse(raw));
                if (record.id === eventID && ['start', 'finish'].includes(String(record.phase))) {
                    jsonOffset(raw, segments);
                    line = index + 1;
                }
            }
            catch { /* Skip incomplete lines during append. */ }
        });
        if (!line)
            throw new Error('event not yet saved');
        return { path: join(directory, id, 'trace.jsonl'), line };
    }
    const run = decodeHistory(raw, id), index = run.events.findIndex(event => event.id === eventID);
    if (index < 0)
        throw new Error('event not found');
    const value = raw.toString(), offset = jsonOffset(value, ['events', String(index), ...segments]);
    return { path: join(directory, id, 'run.json'), line: value.slice(0, offset).split('\n').length };
}
export async function openRecordEditor(location: {
    path: string;
    line: number;
}): Promise<string> {
    const candidates: [
        string,
        string
    ][] = [
        ['/Applications/Cursor.app/Contents/Resources/app/bin/cursor', 'Cursor'],
        ['/Applications/Visual Studio Code.app/Contents/Resources/app/bin/code', 'VS Code'],
        ...(process.env.PATH ?? '').split(delimiter).filter(Boolean).flatMap(folder => [[join(folder, 'cursor'), 'Cursor'], [join(folder, 'code'), 'VS Code']] as [
            string,
            string
        ][]),
    ];
    const candidate = candidates.find(([path]) => { try {
        accessSync(path, constants.X_OK);
        return true;
    }
    catch {
        return false;
    } });
    if (!candidate)
        throw new Error('未找到 Cursor 或 VS Code；请安装编辑器及其 cursor / code 命令');
    await new Promise<void>((resolve, reject) => {
        const child = spawn(candidate[0], ['--goto', `${location.path}:${location.line}`], { stdio: 'ignore', shell: false });
        const timer = setTimeout(() => child.kill('SIGKILL'), 10000);
        child.once('error', error => { clearTimeout(timer); reject(error); });
        child.once('close', code => { clearTimeout(timer); if (code === 0)
            resolve();
        else
            reject(new Error('编辑器未能打开记录；请检查本机 cursor / code 命令')); });
    });
    return candidate[1];
}
